From d095dc74831deee247612bfdec2533666d0aca9a Mon Sep 17 00:00:00 2001
From: Claudio Bley <claudio.bley@tweag.io>
Date: Fri, 30 Jan 2026 15:48:19 +0100
Subject: [PATCH 12/22] Add rpc timeout config and wire it up

By default, each RPC has a timeout of 60 seconds applied to it.
---
 app/buck2_re_configuration/src/lib.rs      |  8 +++++
 remote_execution/oss/re_grpc/src/client.rs | 40 ++++++++++++----------
 2 files changed, 29 insertions(+), 19 deletions(-)

diff --git a/app/buck2_re_configuration/src/lib.rs b/app/buck2_re_configuration/src/lib.rs
index 0d90d3b7b2..49e2bebc3b 100644
--- a/app/buck2_re_configuration/src/lib.rs
+++ b/app/buck2_re_configuration/src/lib.rs
@@ -451,6 +451,8 @@ pub struct Buck2OssReConfiguration {
     pub grpc_keepalive_while_idle: Option<bool>,
     /// Maximum retries for network requests.
     pub max_retries: usize,
+    /// Timeout for RPC requests in seconds. Defaults to 60s.
+    pub rpc_timeout: u64,
 }
 
 #[derive(Clone, Debug, Default, Allocative)]
@@ -574,6 +576,12 @@ impl Buck2OssReConfiguration {
                     property: "max_retries",
                 })?
                 .unwrap_or(0),
+            rpc_timeout: legacy_config
+                .parse(BuckconfigKeyRef {
+                    section: BUCK2_RE_CLIENT_CFG_SECTION,
+                    property: "rpc_timeout",
+                })?
+                .unwrap_or(60),
         })
     }
 }
diff --git a/remote_execution/oss/re_grpc/src/client.rs b/remote_execution/oss/re_grpc/src/client.rs
index f2a7850d91..cb0c93874b 100644
--- a/remote_execution/oss/re_grpc/src/client.rs
+++ b/remote_execution/oss/re_grpc/src/client.rs
@@ -253,6 +253,8 @@ pub struct RERuntimeOpts {
     cas_ttl_secs: i64,
     /// Maximum retries for network requests.
     max_retries: usize,
+    /// Timeout for RPC requests.
+    rpc_timeout: Duration,
 }
 
 #[derive(Clone)]
@@ -345,6 +347,8 @@ impl REClientBuilder {
             http.enforce_http(false);
             let connector = CountingConnector::new(http);
 
+            endpoint = endpoint.timeout(Duration::from_secs(opts.rpc_timeout));
+
             anyhow::Ok(
                 endpoint
                     .connect_with_connector(connector)
@@ -449,6 +453,7 @@ impl REClientBuilder {
                 // on the TTL of the remote blob.
                 cas_ttl_secs: opts.cas_ttl_secs.unwrap_or(60),
                 max_retries: opts.max_retries,
+                rpc_timeout: Duration::from_secs(opts.rpc_timeout),
             },
             grpc_clients,
             capabilities,
@@ -706,7 +711,7 @@ impl REClient {
                             ..Default::default()
                         },
                         metadata,
-                        self.runtime_opts.use_fbcode_metadata,
+                        self.runtime_opts,
                     ))
                     .await?;
 
@@ -744,7 +749,7 @@ impl REClient {
                             ..Default::default()
                         },
                         metadata,
-                        self.runtime_opts.use_fbcode_metadata,
+                        self.runtime_opts,
                     ))
                     .await?;
 
@@ -792,11 +797,7 @@ impl REClient {
                 let metadata = metadata.clone();
 
                 let stream = client
-                    .execute(with_re_metadata(
-                        request,
-                        metadata,
-                        self.runtime_opts.use_fbcode_metadata,
-                    ))
+                    .execute(with_re_metadata(request, metadata, self.runtime_opts))
                     .await?
                     .into_inner();
                 Ok(stream)
@@ -926,7 +927,7 @@ impl REClient {
             |re_request| async {
                 let metadata = metadata.clone();
                 let mut cas_client = self.grpc_clients.cas_client.clone();
-                let use_fbcode_metadata = self.runtime_opts.use_fbcode_metadata;
+                let runtime_opts = self.runtime_opts;
 
                 retry(
                     move || {
@@ -938,7 +939,7 @@ impl REClient {
                                 .batch_update_blobs(with_re_metadata(
                                     re_request,
                                     metadata,
-                                    use_fbcode_metadata,
+                                    runtime_opts,
                                 ))
                                 .await?;
                             Ok(resp.into_inner())
@@ -954,7 +955,7 @@ impl REClient {
             |segments| async {
                 let metadata = metadata.clone();
                 let mut bytestream_client = self.grpc_clients.bytestream_client.clone();
-                let use_fbcode_metadata = self.runtime_opts.use_fbcode_metadata;
+                let runtime_opts = self.runtime_opts;
 
                 retry(
                     move || {
@@ -963,7 +964,7 @@ impl REClient {
                         let requests = futures::stream::iter(segments.clone());
                         async move {
                             let resp = bytestream_client
-                                .write(with_re_metadata(requests, metadata, use_fbcode_metadata))
+                                .write(with_re_metadata(requests, metadata, runtime_opts))
                                 .await?;
 
                             Ok(resp.into_inner())
@@ -1019,7 +1020,7 @@ impl REClient {
             |re_request| async {
                 let metadata = metadata.clone();
                 let mut client = self.grpc_clients.cas_client.clone();
-                let use_fbcode_metadata = self.runtime_opts.use_fbcode_metadata;
+                let runtime_opts = self.runtime_opts;
 
                 retry(
                     move || {
@@ -1031,7 +1032,7 @@ impl REClient {
                                 .batch_read_blobs(with_re_metadata(
                                     re_request,
                                     metadata,
-                                    use_fbcode_metadata,
+                                    runtime_opts,
                                 ))
                                 .await?
                                 .into_inner())
@@ -1046,7 +1047,7 @@ impl REClient {
             },
             |read_request| {
                 let metadata = metadata.clone();
-                let use_fbcode_metadata = self.runtime_opts.use_fbcode_metadata;
+                let runtime_opts = self.runtime_opts;
                 async move {
                     let mut client = self.grpc_clients.bytestream_client.clone();
                     retry(
@@ -1059,7 +1060,7 @@ impl REClient {
                                     .read(with_re_metadata(
                                         read_request,
                                         metadata,
-                                        use_fbcode_metadata,
+                                        runtime_opts,
                                     ))
                                     .await?
                                     .into_inner();
@@ -1119,7 +1120,7 @@ impl REClient {
             // Send a request and notify others of the result
             if !digests_to_check.is_empty() {
                 tracing::debug!(num_digests = digests_to_check.len(), "FindMissingBlobs");
-                let use_fbcode_metadata = self.runtime_opts.use_fbcode_metadata;
+                let runtime_opts = self.runtime_opts;
                 let missing_blobs = retry(
                     || {
                         let mut cas_client = cas_client.clone();
@@ -1137,7 +1138,7 @@ impl REClient {
                                         ..Default::default()
                                     },
                                     metadata,
-                                    use_fbcode_metadata,
+                                    runtime_opts,
                                 ))
                                 .await
                                 .context("Failed to request what blobs are not present on remote")
@@ -1870,7 +1871,7 @@ where
 fn with_re_metadata<T>(
     t: T,
     metadata: RemoteExecutionMetadata,
-    use_fbcode_metadata: bool,
+    runtime_opts: RERuntimeOpts,
 ) -> tonic::Request<T> {
     // This creates a new Tonic request with attached metadata for the RE
     // backend. There are two cases here we need to support:
@@ -1892,8 +1893,9 @@ fn with_re_metadata<T>(
     // Meta builds catch those issues earlier.
 
     let mut msg = tonic::Request::new(t);
+    msg.set_timeout(runtime_opts.rpc_timeout);
 
-    if use_fbcode_metadata {
+    if runtime_opts.use_fbcode_metadata {
         // This is pretty ugly, but the protobuf spec that defines this is
         // internal, so considering field numbers need to be stable anyway (=
         // low risk), and this is not used in prod (= low impact if this goes
-- 
2.51.2

